tests/connect_redirect/connect_redirect_tests.cpp

```diff
@@ -0,0 +1,482 @@
+            v4_proxy = IN6_GET_ADDR_V4MAPPED((IN6_ADDR*)&v6_proxy->sin6_addr);
+        } else {
+            v4_destination = (UCHAR*)&((struct sockaddr_in*)&destination)->sin_addr.S_un.S_addr;
+            v4_proxy = (UCHAR*)&((struct sockaddr_in*)&proxy)->sin_addr.S_un.S_addr;
+        }
+
+        key.destination_ip.ipv4 = *((uint32_t*)v4_destination);
+        value.destination_ip.ipv4 = *((uint32_t*)v4_proxy);
+    } else {
+        struct sockaddr_in6* v6_destination = (struct sockaddr_in6*)&destination;
+        struct sockaddr_in6* v6_proxy = (struct sockaddr_in6*)&proxy;
+        memcpy(key.destination_ip.ipv6, v6_destination->sin6_addr.u.Byte, sizeof(key.destination_ip.ipv6));
+        memcpy(value.destination_ip.ipv6, v6_proxy->sin6_addr.u.Byte, sizeof(value.destination_ip.ipv6));
+    }
+
+    key.destination_port = value.destination_port = htons(port);
+    key.protocol = protocol;
+
+    if (add) {
+        REQUIRE(bpf_map_update_elem(map_fd, &key, &value, 0) == 0);
+    } else {
+        REQUIRE(bpf_map_delete_elem(map_fd, &key) == 0);
+    }
+}
+
+void
+connect_redirect_test(
+    _In_ const struct bpf_object* object,
+    _In_ sender_socket_t* sender_socket,
+    _In_ sockaddr_storage& destination,
+    _In_ sockaddr_storage& proxy,
+    bool dual_stack)
+{
+    // Update policy in the map to redirect the connection to the proxy.
+    _update_policy_map(object, destination, proxy, _globals.remote_port, _globals.protocol, dual_stack, true);
+
+    // Try to send and receive message to "destination". It should succeed.
+    sender_socket->send_message_to_remote_host(CLIENT_MESSAGE, destination, _globals.remote_port);
+    sender_socket->complete_async_send(1000, expected_result_t::success);
+
+    sender_socket->post_async_receive();
+    sender_socket->complete_async_receive(2000, false);
+
+    uint32_t bytes_received = 0;
+    char* received_message = nullptr;
+    sender_socket->get_received_message(bytes_received, received_message);
+
+    // TODO: The message returned by the listener should be unique so that we know
+    // for sure that the connection was indeed redirected.
+    REQUIRE(memcmp(received_message, "test_message", strlen(received_message)) == 0);
+
+    // Remove entry from policy map.
+    _update_policy_map(object, destination, proxy, _globals.remote_port, _globals.protocol, dual_stack, false);
+}
+
+void
+authorize_test(
+    _In_ const struct bpf_object* object,
+    _In_ sender_socket_t* sender_socket,
+    _In_ sockaddr_storage& destination,
+    bool dual_stack)
+{
+    // Default behavior of the eBPF program is to block the connection.
+
+    // Send should fail as the connection is blocked.
+    sender_socket->send_message_to_remote_host(CLIENT_MESSAGE, destination, _globals.remote_port);
+    sender_socket->complete_async_send(1000, expected_result_t::failure);
+
+    // Receive should timeout as connection is blocked.
+    sender_socket->post_async_receive(true);
```

Comment:
Since this is the sender/initiator socket; Isnt it sufficient to check if send fails?

Reply:
In case of UDP, send does not fail, so we have to post a receive.